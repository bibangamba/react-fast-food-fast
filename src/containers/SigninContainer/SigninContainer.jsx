import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { toast } from "react-toastify";
import { signinAction } from "../../actions/signinAction";
import SigninForm from "../../components/Signin/Signin";

export class SigninContainer extends Component {
  state = {
    email: "",
    password: "",
    signinFailed: undefined,
    signinSuccess: undefined,
    signinButtonText: "Signin",
    signinButtonDisabled: false
  };

  componentDidMount() {
    const { error } = this.props;
    if (error) {
      toast.error(error);
    }
  }

  componentWillReceiveProps(nextProps) {
    const { signinFailed, signinSuccess, diffState } = nextProps;
    if (signinFailed) {
      toast.error(signinFailed);
      this.setState({
        signinFailed,
        diffState,
        signinButtonDisabled: false,
        signinButtonText: "Signin"
      });
    } else if (signinSuccess) {
      toast.success(signinSuccess);
      setInterval(this.redirectToOrders, 2000);
    }
  }

  redirectToOrders = () => {
    window.location.replace("/orders");
  }

  onChange = e => this.setState({ [e.target.name]: e.target.value });

  onSubmit = e => {
    e.preventDefault();
    this.setState({
      signinButtonDisabled: true,
      signinButtonText: "Signing in..."
    });
    const { email, password } = this.state;
    const data = {
      email,
      password
    };
    const { signinAction: SignupAction } = this.props;

    SignupAction(data);
  };

  render() {
    const {
      signinSuccess,
      signinFailed,
      diffState,
      signinButtonText,
      signinButtonDisabled
    } = this.state;
    return (
      <SigninForm
        onChange={this.onChange}
        onSubmit={this.onSubmit}
        signinFailed={signinFailed}
        diffState={diffState}
        signinSuccess={signinSuccess}
        signinButtonText={signinButtonText}
        signinButtonDisabled={signinButtonDisabled}
      />
    );
  }
}

SigninContainer.propTypes = {
  signinAction: PropTypes.func.isRequired,
  signinSuccess: PropTypes.string,
  signinFailed: PropTypes.string,
  diffState: PropTypes.string,
  history: PropTypes.shape({}),
  error: PropTypes.string
};

SigninContainer.defaultProps = {
  history: {},
  signinFailed: undefined,
  signinSuccess: undefined,
  diffState: undefined,
  error: undefined
};

export const mapStateToProps = state => ({
  signinSuccess: state.signinReducer.success,
  signinFailed: state.signinReducer.error,
  diffState: state.signinReducer.diff_in_state
});

export default connect(
  mapStateToProps,
  { signinAction }
)(SigninContainer);
