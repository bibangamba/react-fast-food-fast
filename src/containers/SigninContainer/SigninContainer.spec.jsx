import React from "react";
import { shallow } from "enzyme";
import { SigninContainer, mapStateToProps } from "./SigninContainer";

const signinAction = jest.fn();
const changeLoginState = jest.fn();
const props = { signinAction, history: { push: jest.fn() }, changeLoginState, error: "mock error message" };
const initialState = {
  email: "",
  password: "",
  name: "",
  phone: "",
  confirm_password: "",
  signinFailed: undefined,
  signinSuccess: undefined,
  signinReducer: {
    success: "mock success value",
    error: "mock error value"
  },
  signinButtonText: "Signin",
  signinButtonDisabled: false
};

let wrapper;

describe("test signup view", () => {
  beforeEach(() => {
    wrapper = shallow(<SigninContainer {...props} />);
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("test handle onsubmit event", () => {
    wrapper.instance().onSubmit({ preventDefault() { } });
    expect(wrapper.state("signinButtonText")).toEqual("Signing in...");
  });

  it("should handle onchange function", () => {
    wrapper
      .instance()
      .onChange({ target: { name: "email", value: "andrew@g.com" } });
    expect(wrapper.state("email")).toEqual("andrew@g.com");
  });

  it("should receive new fail props", () => {
    wrapper.setProps({
      signinFailed: "mock fail"
    });
    expect(wrapper.state("signinFailed")).toEqual("mock fail");
  });

  it("should receive new success props", () => {
    wrapper.setProps({
      signinSuccess: "mock success",
      signinFailed: undefined
    });
    expect(wrapper.state("signinSuccess")).toEqual(undefined);
  });
  it("should receive new success props", () => {
    wrapper.setProps({
      signinSuccess: undefined,
      signupFail: undefined
    });
    expect(wrapper.state("signinSuccess")).toEqual(undefined);
  });
  it("should map state to props", () => {
    expect(mapStateToProps(initialState).signinSuccess).toEqual(
      "mock success value"
    );
  });
});
