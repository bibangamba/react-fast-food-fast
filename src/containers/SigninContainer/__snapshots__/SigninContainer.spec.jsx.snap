// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test signup view should match the snapshot 1`] = `ShallowWrapper {}`;
